### PR TITLE
Ensure variable offsets point beyond fixed values

### PR DESCRIPTION
**Pull request by kasey**
Fixes https://github.com/ferranbt/fastssz/issues/44 by adding a bounds check to UnmarshalSSZ, in the unmarshaling logic of the first offset in each message. For an example of the new output please see:

https://github.com/prysmaticlabs/ethereumapis/blob/check-variable-offset/eth/v1/generated.ssz.go#L81

### CODE REVIEW

**Comment by ferranbt** [comment]
LGTM!

Could you run 'make build-spec-tests' to regenerate the spec tests? Then we can see the change in the generated code and test that everything works.

**Comment by kasey** [comment]
> LGTM!
> 
> Could you run 'make build-spec-tests' to regenerate the spec tests? Then we can see the change in the generated code and test that everything works.

Hi there, please see new commit for updated spec tests. Also note that the spec test helped me catch an off-by-one error that would have rejected valid encoded values :sweat_smile:

File: spectests/structs_encoding.go [hunk 1]
```diff
@@ -1803,6 +1815,9 @@ func (a *AttesterSlashing) UnmarshalSSZ(buf []byte) error {
 
 	// Field (0) 'Attestation1'
 	{
+		if o0 < 8 {
```
Comment (ferranbt) [inline spectests/structs_encoding.go:1818]: Why there is no offset check for 'Attestation2' in line 1832?

**Review by ferranbt — CHANGES_REQUESTED** [review]

File: sszgen/unmarshal.go [hunk 2]
```diff
@@ -298,8 +301,10 @@ func (v *Value) umarshalContainer(start bool, dst string) (str string) {
 				"name":      i.name,
 				"from":      from,
 				"to":        to,
+				"firstOffsetCheck": firstOffsetCheck(indx),
 				"unmarshal": i.unmarshal("buf"),
 			})
+			firstOffsetCheck = func(int) string { return "" }
```
[the remark this answers is not in the document]
Comment (kasey) [reply sszgen/unmarshal.go:307]: Correct, it is only inserting the check for the first offset.

File: spectests/structs_encoding.go (hunk 1 above)
Comment (kasey) [reply spectests/structs_encoding.go:1818]: My thinking was that after the first check, the second check incorporates the first offset as it's beginning slice, so all slice offsets are aligned to each other. If any offset after the first is lower than the previous offset, the go runtime will panic. We could explicitly check so that bad values can be rejected without causing a panic.

Comment (kasey) [reply spectests/structs_encoding.go:1818]: Yeah on second thought I like explicit checks to avoid a runtime panic, I'm rewriting this code to check each offset now, thanks!

File: spectests/structs_encoding.go [hunk 3]
```diff
@@ -913,6 +922,9 @@ func (p *PendingAttestation) UnmarshalSSZ(buf []byte) error {
 
 	// Field (0) 'AggregationBits'
 	{
+		if o0 < 148 {
```
[the remark this answers is not in the document]
Comment (kasey) [reply spectests/structs_encoding.go:925]: sounds good

[the remark this answers is not in the document]
Comment (kasey) [reply spectests/structs_encoding.go:925]: To be clear, I interpreted "in `ReadOffset`" as "where we call `ReadOffset`", not actually *in* `ReadOffset`, as `ReadOffset` simply decodes the offset value from the byte slice with no bounds checking, and adding bounds checking to it would clutter its method signature (it would need to know about boundary offsets to perform checks).

**Comment by kasey** [comment]
Hey @ferranbt good suggestions, updated based on your input, please take another look!

File: spectests/structs_encoding.go [hunk 4]
```diff
@@ -1795,11 +1807,17 @@ func (a *AttesterSlashing) UnmarshalSSZ(buf []byte) error {
...
 	}
+	if o0 < 8 {
+		return ssz.ErrInvalidVariableOffset
+	}
 
 	// Offset (1) 'Attestation2'
 	if o1 = ssz.ReadOffset(buf[4:8]); o1 > size || o0 > o1 {
 		return ssz.ErrOffset
 	}
+	if o1 < o0 {
```
Comment (ferranbt) [inline spectests/structs_encoding.go:1818]: This check is not necessary since its already done in line 1815.

File: spectests/structs_encoding.go [hunk 5]
```diff
@@ -1795,11 +1807,17 @@ func (a *AttesterSlashing) UnmarshalSSZ(buf []byte) error {
 	if o0 = ssz.ReadOffset(buf[0:4]); o0 > size {
 		return ssz.ErrOffset
 	}
+	if o0 < 8 {
```
Comment (ferranbt) [inline spectests/structs_encoding.go:1810]: Can you merge this check with the one in line 1807? Both are part of the same offset check

Comment (ferranbt) [inline spectests/structs_encoding.go:1810]: Shouldn't it be?
```
if o0 != 8 {
```

File: sszgen/unmarshal.go [hunk 6]
```diff
@@ -268,26 +274,29 @@ func (v *Value) umarshalContainer(start bool, dst string) (str string) {
 			if {{.offset}} = ssz.ReadOffset({{.dst}}); {{.offset}} > size {{.more}} {
 				return ssz.ErrOffset
 			}
+			if {{.offset}} < {{.offsetMinimum}} {
```
Comment (ferranbt) [inline sszgen/unmarshal.go:277]: You can reuse the "more" variable to include the first offset check. Something like this should work:
```
if prev, ok := offsetsMatch[offset]; ok {
    // check that the offsets are higher than the previous one
    data["more"] = fmt.Sprintf(" || %s > %s", prev, offset)
} else {
   // check that the first offset starts at the correct position
   data["more"] = fmt.Sprintf(" || %s > %d", offset, v.n)
}
```

**Review by ferranbt — CHANGES_REQUESTED** [review]
Hi! Sorry because I gave you some mixed answers before. Now I fully understand the goal of the PR. I added the review and should be ready to merge then

File: spectests/structs_encoding.go (hunk 5 above)
Comment (ferranbt) [reply spectests/structs_encoding.go:1810]: See the comment in unmarshal.go for a possible solution

**Comment by kasey** [comment]
Teku and Lighthouse both specify a separate error case for a mismatch of the alignment of the first variable-length offset. This logic was designed to match those implementations. I think having these more fine-grained errors is helpful in understanding what the encoding issue is. I also think packing more cases into the `more` variable would hurt the readability of the generated code and the template. The existing code already has a `>` check, so adding a `<` explicitly checks for both kinds of inequality and together are equivalent to `!=`. If you don't mind I would like to preserve the current behavior for the first offset. I'll update the code to remove the check for subsequent offsets, because those are redundant (and therefore unreachable). What do you think?

Here are the lighthouse and teku implementations for reference:
https://github.com/sigp/lighthouse/blob/stable/consensus/ssz/src/decode.rs#L76
https://github.com/ConsenSys/teku/blob/9c0086ca5af0f90eddc26ca437475adfa7afc914/ssz/src/main/java/tech/pegasys/teku/ssz/schema/impl/AbstractSszContainerSchema.java#L278

**Comment by kasey** [comment]
@ferranbt please take a look at the new commit. I tried to address all your per-line feedback in my previous comment on the PR. Thank you!

**Comment by ferranbt** [comment]
Ok, sounds perfect to me. I will merge it right away!
Thank you for the PR.

**Review by ferranbt — APPROVED** [review]

**Comment by kasey** [comment]
Thank you!
